File: mobile/lib/models/candidate.dart
/// CandidateStatus — matches backend candidates.status values exactly.
enum CandidateStatus {
  draft,
  kycPending,
  addressPending,
  documentPending,
  documentVerificationPending,
  biometricPending,
  completed,
  rejected,
}

const Map<String, CandidateStatus> _statusFromString = {
  'DRAFT': CandidateStatus.draft,
  'KYC_PENDING': CandidateStatus.kycPending,
  'ADDRESS_PENDING': CandidateStatus.addressPending,
  'DOCUMENT_PENDING': CandidateStatus.documentPending,
  'DOCUMENT_VERIFICATION_PENDING': CandidateStatus.documentVerificationPending,
  'BIOMETRIC_PENDING': CandidateStatus.biometricPending,
  'COMPLETED': CandidateStatus.completed,
  'REJECTED': CandidateStatus.rejected,
};

const Map<CandidateStatus, String> _statusToString = {
  CandidateStatus.draft: 'DRAFT',
  CandidateStatus.kycPending: 'KYC_PENDING',
  CandidateStatus.addressPending: 'ADDRESS_PENDING',
  CandidateStatus.documentPending: 'DOCUMENT_PENDING',
  CandidateStatus.documentVerificationPending: 'DOCUMENT_VERIFICATION_PENDING',
  CandidateStatus.biometricPending: 'BIOMETRIC_PENDING',
  CandidateStatus.completed: 'COMPLETED',
  CandidateStatus.rejected: 'REJECTED',
};

CandidateStatus candidateStatusFromString(String value) =>
    _statusFromString[value] ?? CandidateStatus.draft;

String candidateStatusToString(CandidateStatus status) => _statusToString[status]!;

String candidateStatusLabel(CandidateStatus status) {
  switch (status) {
    case CandidateStatus.draft:
      return 'Draft';
    case CandidateStatus.kycPending:
      return 'KYC Pending';
    case CandidateStatus.addressPending:
      return 'Address Pending';
    case CandidateStatus.documentPending:
      return 'Document Pending';
    case CandidateStatus.documentVerificationPending:
      return 'Doc. Verification Pending';
    case CandidateStatus.biometricPending:
      return 'Biometric Pending';
    case CandidateStatus.completed:
      return 'Completed';
    case CandidateStatus.rejected:
      return 'Rejected';
  }
}

/// Server-side progress marker (candidates.current_step). Records the step
/// just COMPLETED — resume-on-reopen opens the NEXT step (see
/// core/wizard_steps.dart RESUME_STEP map).
enum CandidateStep {
  registration,
  kyc,
  address,
  documentChecklist,
  documentUpload,
  verification,
  originalVerification,
  photo,
  leftBiometric,
  rightBiometric,
  declaration,
  signature,
  submitted,
}

const Map<String, CandidateStep> _stepFromString = {
  'REGISTRATION': CandidateStep.registration,
  'KYC': CandidateStep.kyc,
  'ADDRESS': CandidateStep.address,
  'DOCUMENT_CHECKLIST': CandidateStep.documentChecklist,
  'DOCUMENT_UPLOAD': CandidateStep.documentUpload,
  'VERIFICATION': CandidateStep.verification,
  'ORIGINAL_VERIFICATION': CandidateStep.originalVerification,
  'PHOTO': CandidateStep.photo,
  'LEFT_BIOMETRIC': CandidateStep.leftBiometric,
  'RIGHT_BIOMETRIC': CandidateStep.rightBiometric,
  'DECLARATION': CandidateStep.declaration,
  'SIGNATURE': CandidateStep.signature,
  'SUBMITTED': CandidateStep.submitted,
};

const Map<CandidateStep, String> _stepToString = {
  CandidateStep.registration: 'REGISTRATION',
  CandidateStep.kyc: 'KYC',
  CandidateStep.address: 'ADDRESS',
  CandidateStep.documentChecklist: 'DOCUMENT_CHECKLIST',
  CandidateStep.documentUpload: 'DOCUMENT_UPLOAD',
  CandidateStep.verification: 'VERIFICATION',
  CandidateStep.originalVerification: 'ORIGINAL_VERIFICATION',
  CandidateStep.photo: 'PHOTO',
  CandidateStep.leftBiometric: 'LEFT_BIOMETRIC',
  CandidateStep.rightBiometric: 'RIGHT_BIOMETRIC',
  CandidateStep.declaration: 'DECLARATION',
  CandidateStep.signature: 'SIGNATURE',
  CandidateStep.submitted: 'SUBMITTED',
};

CandidateStep candidateStepFromString(String value) =>
    _stepFromString[value] ?? CandidateStep.registration;

String candidateStepToString(CandidateStep step) => _stepToString[step]!;

enum Gender { male, female, other }

Gender? genderFromString(String? value) {
  switch (value) {
    case 'MALE':
      return Gender.male;
    case 'FEMALE':
      return Gender.female;
    case 'OTHER':
      return Gender.other;
  }
  return null;
}

String genderToString(Gender gender) {
  switch (gender) {
    case Gender.male:
      return 'MALE';
    case Gender.female:
      return 'FEMALE';
    case Gender.other:
      return 'OTHER';
  }
}

class Candidate {
  final int id;
  final String candidateNumber;
  final String fullName;
  final String mobile;
  final String? email;
  final String? dob;
  final Gender? gender;
  final int? coordinatorId;
  final String? coordinatorName;
  final CandidateStatus status;
  final CandidateStep currentStep;
  final String? submittedAt;
  final String createdAt;
  final String updatedAt;

  Candidate({
    required this.id,
    required this.candidateNumber,
    required this.fullName,
    required this.mobile,
    this.email,
    this.dob,
    this.gender,
    this.coordinatorId,
    this.coordinatorName,
    required this.status,
    required this.currentStep,
    this.submittedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Candidate.fromJson(Map<String, dynamic> json) {
    return Candidate(
      id: json['id'] as int,
      candidateNumber: json['candidate_number'] as String,
      fullName: json['full_name'] as String,
      mobile: json['mobile'] as String,
      email: json['email'] as String?,
      dob: json['dob'] as String?,
      gender: genderFromString(json['gender'] as String?),
      coordinatorId: json['coordinator_id'] as int?,
      coordinatorName: json['coordinator_name'] as String?,
      status: candidateStatusFromString(json['status'] as String),
      currentStep: candidateStepFromString(json['current_step'] as String),
      submittedAt: json['submitted_at'] as String?,
      createdAt: json['created_at']?.toString() ?? '',
      updatedAt: json['updated_at']?.toString() ?? '',
    );
  }
}

/// Request payload for POST /candidates.
class CandidateRegistrationPayload {
  final String fullName;
  final String mobile;
  final String? email;
  final String? dob;
  final Gender? gender;
  final int? coordinatorId;

  CandidateRegistrationPayload({
    required this.fullName,
    required this.mobile,
    this.email,
    this.dob,
    this.gender,
    this.coordinatorId,
  });

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{'fullName': fullName, 'mobile': mobile};
    if (email != null && email!.isNotEmpty) map['email'] = email;
    if (dob != null && dob!.isNotEmpty) map['dob'] = dob;
    if (gender != null) map['gender'] = genderToString(gender!);
    if (coordinatorId != null) map['coordinatorId'] = coordinatorId;
    return map;
  }
}

class StatusHistoryEntry {
  final int id;
  final int candidateId;
  final String? oldStatus;
  final String newStatus;
  final int? changedBy;
  final String? remarks;
  final String createdAt;

  StatusHistoryEntry({
    required this.id,
    required this.candidateId,
    this.oldStatus,
    required this.newStatus,
    this.changedBy,
    this.remarks,
    required this.createdAt,
  });

  factory StatusHistoryEntry.fromJson(Map<String, dynamic> json) {
    return StatusHistoryEntry(
      id: json['id'] as int,
      candidateId: json['candidate_id'] as int,
      oldStatus: json['old_status'] as String?,
      newStatus: json['new_status'] as String,
      changedBy: json['changed_by'] as int?,
      remarks: json['remarks'] as String?,
      createdAt: json['created_at']?.toString() ?? '',
    );
  }
}

File: frontend/src/types/candidate.ts
export type CandidateStatus =
  | 'DRAFT'
  | 'KYC_PENDING'
  | 'ADDRESS_PENDING'
  | 'DOCUMENT_PENDING'
  | 'DOCUMENT_VERIFICATION_PENDING'
  | 'BIOMETRIC_PENDING'
  | 'COMPLETED'
  | 'REJECTED';

export const CANDIDATE_STATUSES: CandidateStatus[] = [
  'DRAFT',
  'KYC_PENDING',
  'ADDRESS_PENDING',
  'DOCUMENT_PENDING',
  'DOCUMENT_VERIFICATION_PENDING',
  'BIOMETRIC_PENDING',
  'COMPLETED',
  'REJECTED',
];

/** Server-side progress marker (candidates.current_step). */
export type CandidateStep =
  | 'REGISTRATION'
  | 'KYC'
  | 'ADDRESS'
  | 'DOCUMENT_CHECKLIST'
  | 'DOCUMENT_UPLOAD'
  | 'VERIFICATION'
  | 'ORIGINAL_VERIFICATION'
  | 'PHOTO'
  | 'LEFT_BIOMETRIC'
  | 'RIGHT_BIOMETRIC'
  | 'DECLARATION'
  | 'SIGNATURE'
  | 'SUBMITTED';

export type Gender = 'MALE' | 'FEMALE' | 'OTHER';

export interface Candidate {
  id: number;
  candidate_number: string;
  full_name: string;
  mobile: string;
  email: string | null;
  dob: string | null;
  gender: Gender | null;
  coordinator_id: number | null;
  /** Joined from the coordinators table on list/detail queries. */
  coordinator_name?: string | null;
  status: CandidateStatus;
  current_step: CandidateStep;
  submitted_at: string | null;
  created_by: number | null;
  created_at: string;
  updated_at: string;
}

/** Paginated list envelope used by /candidates and /audit-logs. */
export interface PaginatedResult<T> {
  rows: T[];
  total: number;
  page?: number;
  limit?: number;
}

/** Request payload for POST /candidates. */
export interface CandidateRegistrationPayload {
  fullName: string;
  mobile: string;
  email?: string;
  dob?: string;
  gender?: Gender;
  coordinatorId?: number | string;
}

export type KycGender = 'MALE' | 'FEMALE';
export type MaritalStatus = 'SINGLE' | 'MARRIED';
export type ResidencyStatus = 'RESIDENT_INDIVIDUAL' | 'NON_RESIDENT' | 'FOREIGN_NATIONAL';
export type ProofOfIdentity = 'AADHAAR' | 'PAN' | 'PASSPORT' | 'DL' | 'VOTER_ID' | 'OTHER';

/** GET /candidates/:id/kyc — identity numbers come back masked. */
export interface KycData {
  id: number;
  candidate_id: number;
  applicant_name: string | null;
  father_spouse_name: string | null;
  gender: KycGender | null;
  marital_status: MaritalStatus | null;
  dob: string | null;
  nationality: string;
  residency_status: ResidencyStatus | null;
  pan_number: string | null;
  aadhaar_number: string | null;
  proof_of_identity: ProofOfIdentity | null;
  kyc_provider_mode: string;
  kyc_reference: string | null;
  is_completed: 0 | 1 | boolean;
  created_at: string;
  updated_at: string;
}

/** PUT /candidates/:id/kyc payload. PAN/Aadhaar are omitted when left blank. */
export interface KycPayload {
  applicantName: string;
  fatherSpouseName?: string;
  gender: KycGender | '';
  maritalStatus?: MaritalStatus | '';
  dob: string;
  nationality?: string;
  residencyStatus?: ResidencyStatus;
  panNumber?: string;
  aadhaarNumber?: string;
  proofOfIdentity: ProofOfIdentity | '';
}

export type ProofOfAddress =
  | 'AADHAAR'
  | 'PASSPORT'
  | 'UTILITY_BILL'
  | 'BANK_STATEMENT'
  | 'RENT_AGREEMENT'
  | 'OTHER';

/** GET /candidates/:id/address */
export interface AddressData {
  id: number;
  candidate_id: number;
  residence_address: string | null;
  residence_pin_code: string | null;
  contact_email: string | null;
  contact_mobile: string | null;
  proof_of_address: ProofOfAddress | null;
  same_as_residence: 0 | 1 | boolean;
  permanent_address: string | null;
  permanent_pin_code: string | null;
  is_completed: 0 | 1 | boolean;
  created_at: string;
  updated_at: string;
}

/** PUT /candidates/:id/address payload. */
export interface AddressPayload {
  residenceAddress: string;
  residencePinCode: string;
  contactEmail?: string;
  contactMobile?: string;
  proofOfAddress: ProofOfAddress | '';
  sameAsResidence: boolean;
  permanentAddress?: string;
  permanentPinCode?: string;
}

/** GET/PUT /candidates/:id/original-verification */
export interface OriginalVerification {
  id: number;
  candidate_id: number;
  originals_verified: 0 | 1 | boolean;
  self_attested_received: 0 | 1 | boolean;
  verified_by: number | null;
  verified_at: string | null;
  created_at: string;
  updated_at: string;
}

/** GET/PUT /candidates/:id/declaration */
export interface Declaration {
  id: number;
  candidate_id: number;
  accepted: 0 | 1 | boolean;
  accepted_at: string | null;
  ip_address: string | null;
  created_at: string;
  updated_at: string;
}

export type BiometricHand = 'LEFT_HAND' | 'RIGHT_HAND';
export type BiometricVerificationStatus = 'CAPTURED' | 'VERIFIED' | 'FAILED';

/** GET /candidates/:id/biometric — no raw fingerprint data, only references + scores. */
export interface BiometricRecord {
  id: number;
  candidate_id: number;
  hand: BiometricHand;
  provider: string;
  capture_reference: string;
  /** DECIMAL(5,2) — MySQL driver may hand this back as a string. */
  quality_score: number | string;
  device_info: string | null;
  verification_status: BiometricVerificationStatus;
  captured_by: number | null;
  captured_at: string;
  created_at: string;
}

/** GET /biometric/device-status */
export interface BiometricDeviceStatus {
  provider: string;
  connected: boolean;
  deviceName: string;
}
